staging/src/k8s.io/apiserver/pkg/server/storage/resource_config.go

```diff
@@ -105,7 +105,11 @@ func (o *ResourceConfig) EnableVersions(versions ...schema.GroupVersion) {
 
 // TODO this must be removed and we enable/disable individual resources.
 func (o *ResourceConfig) versionEnabled(version schema.GroupVersion) bool {
-	enabled, _ := o.GroupVersionConfigs[version]
+	enabled, ok := o.GroupVersionConfigs[version]
+	if !ok {
+		// enabled by default
```

Comment:
I'm not familiar with this field to tell if this is the actual behavior, and there is [no documentation](https://github.com/kubernetes/kubernetes/blob/2de5cb9abcbcb0d793dcd74627b20019d9b9e386/staging/src/k8s.io/apiserver/pkg/server/storage/resource_config.go#L32:2) about this. Could you share an example of this behavior?

Reply:
Sure, without this change- when I am trying to check if a resource is enabled on an apiserver using [ResourceEnabled()](https://github.com/kubernetes/kubernetes/blob/release-1.27/staging/src/k8s.io/apiserver/pkg/server/storage/resource_config.go#L124-L136), the `versionEnabled()` was returning false for the following resources:

```
"admissionregistration.k8s.io" 
"apiextensions.k8s.io" 
"apiregistration.k8s.io" 
"apps" 
"autoscaling" 
"batch" 
"certificates.k8s.io" 
"coordination.k8s.io" 
"discovery.k8s.io" 
"flowcontrol.apiserver.k8s.io" 
"internal.apiserver.k8s.io" 
"networking.k8s.io" 
"node.k8s.io" 
```

and the GroupVersionConfigs map was reporting just these 2 keys:

- ["apiregistration.k8s.io" , "v1"] : true
- ["apiregistration.k8s.io", "v1beta1"] : true

Reply:
I'm not sure if we can assume the "enabled by default" behavior and it seems unsafe to change. 

What you were seeing seems to suggest that you were querying the aggregator server, which serves the `apiregistration.k8s.io` API. IIRC the other groups are served by the core kube-apiserver (most of the groups) and the apiextensions-apiserver (apiextensions.k8s.io). Could you check that?

Maybe it would be great to understand your use case and what code path you were hitting. Could you elaborate?

Reply:
So basically the changes that I made in [this](https://github.com/kubernetes/kubernetes/pull/118386) PR are not working, since I was trying to add "served versions" info in storageversion api. To find these served versions I was using ResourceEnabled() [here](https://github.com/kubernetes/kubernetes/pull/118386/files#diff-72d684ebcd165e431951adc7c1912c1f615ddf5d14f2a2192026094be271febeR150) which always returns false for the listed groups. So ServedVersions are always coming out to be an empty list.

Because of this when I try to use these served versions info in my UVIP PR [here](https://github.com/kubernetes/kubernetes/pull/117740/files#diff-7bcd0a60c323058ec64ca92eaad9b056ab18c0178f8056fee19d9302974748b9R371) (right now it says DecodableVersions, but plan is to use ServedVersions) I get back empty result and hence requests to these resources fail.

Reply:
It does sound like the comment in the calling code was incorrect and the `MergedResourceConfig` is not the one we expect. We should avoid enable by default given how frequently we're actually disabled by default.

Reply:
Then is using a loopback client to find the list of enabled resources a way out here?

Reply:
Btw ResourceEnabled() does eventually assume the resource to be enabled by default https://github.com/Richabanker/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/storage/resource_config.go#L134C1-L135 
it seems wrong to use the default bool value (false) when seen that the key in the map doesn't exist. Are we sure this is in fact the intended behavior?

Reply:
Resources inside of a version are different than versions.  We want the versions to be disabled, but unknown resources inside of enabled versions are enabled.

so if we have
* group/known-version-1=enabled
* group/known-version-2=disabled
* group/known-version-2/resource-A=enabled

then check
1. group/unknown-version/resource-A enabled?  no
2. group/known-version-2/resource-A enabled?  yes
3. group/known-version-2/resource-B enabled? no
4. group/known-version-1/resource-A enabled? yes
5. group/known-version-2/resource-B enabled? yes

Reply:
I see, yeah that's what the ResourceEnabled() is doing then. 
What I'm seeing is for the kube-aggregator server, the GroupVersionConfigs map has just 2 keys (both for apiregistration with different versions) and ResourceConfigs map is empty.

But I do see, some 30+ entries also in another GroupVersionConfigs map at some point, that's probably for the generic apiserver? These are the entries I see 
```
map[v1:true admissionregistration.k8s.io/v1:true admissionregistration.k8s.io/v1alpha1:true apps/v1:true authentication.k8s.io/v1:true authentication.k8s.io/v1alpha1:true authentication.k8s.io/v1beta1:true authorization.k8s.io/v1:true autoscaling/v1:true autoscaling/v2:true batch/v1:false certificates.k8s.io/v1:true certificates.k8s.io/v1alpha1:true coordination.k8s.io/v1:true discovery.k8s.io/v1:true events.k8s.io/v1:true flowcontrol.apiserver.k8s.io/v1alpha1:true flowcontrol.apiserver.k8s.io/v1beta1:true flowcontrol.apiserver.k8s.io/v1beta2:true flowcontrol.apiserver.k8s.io/v1beta3:true internal.apiserver.k8s.io/v1alpha1:true networking.k8s.io/v1:true networking.k8s.io/v1alpha1:true node.k8s.io/v1:true policy/v1:true policy/v1beta1:true rbac.authorization.k8s.io/v1:true resource.k8s.io/v1alpha2:true scheduling.k8s.io/v1:true storage.k8s.io/v1:true storage.k8s.io/v1alpha1:true storage.k8s.io/v1beta1:true]
```

Reply:
I also did some instrumentation locally and see maps containing the other groups. I think that supports the theory of there being multiple servers (aggregator, core, apiextensions) in the delegation chain, each holding information of the APIs they serve
```
map[v1:true admissionregistration.k8s.io/v1:true admissionregistration.k8s.io/v1alpha1:false apps/v1:true authentication.k8s.io/v1:true authentication.k8s.io/v1alpha1:false authentication.k8s.io/v1beta1:false authorization.k8s.io/v1:true autoscaling/v1:true autoscaling/v2:true batch/v1:true certificates.k8s.io/v1:true certificates.k8s.io/v1alpha1:false coordination.k8s.io/v1:true discovery.k8s.io/v1:true events.k8s.io/v1:true flowcontrol.apiserver.k8s.io/v1alpha1:false flowcontrol.apiserver.k8s.io/v1beta1:false flowcontrol.apiserver.k8s.io/v1beta2:false flowcontrol.apiserver.k8s.io/v1beta3:false internal.apiserver.k8s.io/v1alpha1:false networking.k8s.io/v1:true networking.k8s.io/v1alpha1:false node.k8s.io/v1:true policy/v1:true policy/v1beta1:false rbac.authorization.k8s.io/v1:true resource.k8s.io/v1alpha2:false scheduling.k8s.io/v1:true storage.k8s.io/v1:true storage.k8s.io/v1alpha1:false storage.k8s.io/v1beta1:false]
```
```
map[apiextensions.k8s.io/v1:true apiextensions.k8s.io/v1beta1:true]
```